File: backend/app/schemas/comparison.py
from datetime import datetime
from typing import List, Optional, Literal

from pydantic import BaseModel, Field, UUID4


COMPARISON_DISCLAIMER = (
    "Comparison data comes from the Padhaanewala verified database only. "
    "Fees, cutoffs and admission status change by year - verify current details "
    "with each institution before deciding."
)


class ComparisonRequest(BaseModel):
    college_ids: List[UUID4] = Field(..., min_length=1, max_length=4)
    course_id: Optional[UUID4] = Field(None, description="Optional focus course")


class ComparisonCourse(BaseModel):
    course_id: UUID4
    name: str
    level: Optional[str] = None
    duration_months: Optional[int] = None
    fees: Optional[float] = None
    intake: Optional[int] = None


class ComparisonCutoff(BaseModel):
    course_name: str
    exam_name: Optional[str] = None
    year: int
    category: Optional[str] = None
    opening_rank: Optional[int] = None
    closing_rank: Optional[int] = None


class ComparisonReview(BaseModel):
    id: UUID4
    rating: int
    title: Optional[str] = None
    content: Optional[str] = None
    created_at: datetime


class ComparisonCollege(BaseModel):
    id: UUID4
    name: str
    slug: str
    official_name: Optional[str] = None
    college_type: Optional[str] = None
    is_private: bool = True
    accreditation: Optional[str] = None
    recognition: Optional[str] = None
    established_year: Optional[int] = None
    university_name: Optional[str] = None
    state: Optional[str] = None
    district: Optional[str] = None
    city: Optional[str] = None
    pincode: Optional[str] = None
    address: Optional[str] = None
    website: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    entrance_exam: Optional[str] = None
    admission_status: Optional[str] = None
    has_hostel: bool = False
    rating: Optional[float] = None
    courses: List[ComparisonCourse] = []
    cutoffs: List[ComparisonCutoff] = []
    facilities: List[str] = []
    reviews: List[ComparisonReview] = []
    eligibility: Optional[str] = None
    admission_process: Optional[str] = None
    verification_status: Optional[str] = None
    source_name: Optional[str] = None
    last_verified_at: Optional[datetime] = None


class ComparisonResponse(BaseModel):
    colleges: List[ComparisonCollege]
    course_id: Optional[UUID4] = None
    disclaimer: str = COMPARISON_DISCLAIMER


# ---- "Ask AI: Which college is better for me?" ----
# Deterministic, rules-based analysis over verified DB fields only. The output
# contract is stable so an LLM service can replace the scorer later (Phase 26)
# without changing the frontend.

class ComparisonPreferences(BaseModel):
    course: Optional[str] = None
    budget: Optional[float] = Field(None, ge=0)
    requires_hostel: Optional[bool] = None
    prefers_govt: Optional[bool] = None
    state: Optional[str] = None
    city: Optional[str] = None


class AICompareRequest(BaseModel):
    college_ids: List[UUID4] = Field(..., min_length=1, max_length=4)
    preferences: Optional[ComparisonPreferences] = Field(
        default_factory=ComparisonPreferences
    )


class AICollegeAnalysis(BaseModel):
    college_id: UUID4
    name: str
    slug: str
    tier: Literal["HIGHLY_SUITABLE", "POSSIBLE", "REACH"]
    score: int = Field(..., ge=0, le=100)
    summary: str
    strengths: List[str] = []
    weaknesses: List[str] = []
    sources: List[str] = []


class AICompareResponse(BaseModel):
    colleges: List[AICollegeAnalysis]
    overall_summary: str
    disclaimer: str = (
        "This is an automated estimate based only on Padhaanewala's verified database. "
        "It helps you shortlist colleges - it is NOT an admission guarantee and does "
        "not replace official eligibility and admission rules."
    )

File: frontend/src/types/comparison.ts
// Types mirroring backend schemas/comparison.py.

export interface ComparisonCourse {
  course_id: string
  name: string
  level?: string | null
  duration_months?: number | null
  fees?: number | null
  intake?: number | null
}

export interface ComparisonCutoff {
  course_name: string
  exam_name?: string | null
  year: number
  category?: string | null
  opening_rank?: number | null
  closing_rank?: number | null
}

export interface ComparisonReview {
  id: string
  rating: number
  title?: string | null
  content?: string | null
  created_at: string
}

export interface ComparisonCollege {
  id: string
  name: string
  slug: string
  official_name?: string | null
  college_type?: string | null
  is_private: boolean
  accreditation?: string | null
  recognition?: string | null
  established_year?: number | null
  university_name?: string | null
  state?: string | null
  district?: string | null
  city?: string | null
  pincode?: string | null
  address?: string | null
  website?: string | null
  email?: string | null
  phone?: string | null
  entrance_exam?: string | null
  admission_status?: string | null
  has_hostel: boolean
  rating?: number | null
  courses: ComparisonCourse[]
  cutoffs: ComparisonCutoff[]
  facilities: string[]
  reviews: ComparisonReview[]
  eligibility?: string | null
  admission_process?: string | null
  verification_status?: string | null
  source_name?: string | null
  last_verified_at?: string | null
}

export interface ComparisonResponse {
  colleges: ComparisonCollege[]
  course_id?: string | null
  disclaimer: string
}

export interface ComparisonPreferences {
  course?: string
  budget?: number
  requires_hostel?: boolean
  prefers_govt?: boolean
  state?: string
  city?: string
}

export type AiTier = "HIGHLY_SUITABLE" | "POSSIBLE" | "REACH"

export interface AiCollegeAnalysis {
  college_id: string
  name: string
  slug: string
  tier: AiTier
  score: number
  summary: string
  strengths: string[]
  weaknesses: string[]
  sources: string[]
}

export interface AiCompareResponse {
  colleges: AiCollegeAnalysis[]
  overall_summary: string
  disclaimer: string
}